File: src/ZoomacIt/Overlay/BreakTimerRingGeometry.swift
import CoreGraphics
import Foundation

/// Pure math for the Break Timer widget's progress ring and expiration pulse. No
/// AppKit dependency — `BreakTimerView` is the only consumer, and keeping this here
/// (rather than inline in `draw(_:)`) makes both formulas unit-testable.
enum BreakTimerRingGeometry {

    /// Fraction of the ring to draw as "time remaining" (1.0 = full circle, 0.0 = empty).
    static func remainingFraction(remainingSeconds: Int, totalSeconds: Int) -> CGFloat {
        guard totalSeconds > 0 else { return 0 }
        let fraction = CGFloat(remainingSeconds) / CGFloat(totalSeconds)
        return min(max(fraction, 0), 1)
    }

    /// Alpha for the pulsing red ring while expired, oscillating between 0.35 and 1.0
    /// on a 0.9s cycle. `elapsedTime` is any monotonically increasing clock (the
    /// caller uses `CACurrentMediaTime()`).
    static func expiredPulseAlpha(elapsedTime: TimeInterval) -> CGFloat {
        let period: TimeInterval = 0.9
        let phase = elapsedTime.truncatingRemainder(dividingBy: period) / period
        let wave = (sin(phase * 2 * .pi) + 1) / 2 // 0...1
        return 0.35 + wave * 0.65 // 0.35...1.0
    }
}

File: windows/Overlay/BreakTimerRingGeometry.cs
namespace ZoomacItWin.Overlay;

/// <summary>
/// 진행 링과 만료 깜빡임의 순수 계산.
/// (맥 원본 <c>Overlay/BreakTimerRingGeometry.swift</c>의 직역 이식)
/// </summary>
internal static class BreakTimerRingGeometry
{
    /// <summary>"남은 시간"으로 그릴 링의 비율 (1.0 = 원 전체, 0.0 = 없음).</summary>
    public static double RemainingFraction(int remainingSeconds, int totalSeconds)
    {
        if (totalSeconds <= 0) return 0;
        double fraction = (double)remainingSeconds / totalSeconds;
        return Math.Min(Math.Max(fraction, 0), 1);
    }

    /// <summary>
    /// 만료 후 빨간 링이 깜빡일 때의 불투명도. 0.9초 주기로 0.35 ~ 1.0 사이를 오간다.
    /// <paramref name="elapsedTime"/>은 단조 증가하는 아무 시계나 된다.
    /// </summary>
    public static double ExpiredPulseAlpha(double elapsedTime)
    {
        const double period = 0.9;
        double phase = (elapsedTime % period) / period;
        double wave = (Math.Sin(phase * 2 * Math.PI) + 1) / 2; // 0...1
        return 0.35 + wave * 0.65;                             // 0.35...1.0
    }
}
